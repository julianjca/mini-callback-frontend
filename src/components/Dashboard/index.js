import React from 'react'
import { Flex, Box, Button, Heading, Table, Thead, Tbody, Tr, Th, Td } from '@chakra-ui/react'

const Dashboard = ({ callbacks }) => {
  return (
    <Box pt={20} pb={20}>
      <Flex alignItems="center" justifyContent="space-between">
        <Box>
          <Heading>Dashboard</Heading>
        </Box>
        <Box>
          <Button onClick={() => console.log('log out')} size="lg" width={200} colorScheme="blue">
            Log out
          </Button>
        </Box>
      </Flex>
      <Table mt={20} variant="simple">
        <Thead>
          <Tr>
            <Th>Callback ID</Th>
            <Th>Transaction ID</Th>
            <Th>Virtual Account</Th>
            <Th>Bank Code</Th>
            <Th>Detail</Th>
          </Tr>
        </Thead>
        {callbacks.map(callback => (
          <Tbody key={callback.id}>
            <Tr>
              <Td>{callback.id}</Td>
              <Td>{callback.transaction_id}</Td>
              <Td>{callback.virtual_account}</Td>
              <Td>{callback.bank_code}</Td>
              <Td>
                <Button size="sm">See detail</Button>
              </Td>
            </Tr>
          </Tbody>
        ))}
      </Table>
    </Box>
  )
}

export default Dashboard
