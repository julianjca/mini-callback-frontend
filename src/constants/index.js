export const SET_LOGIN_STATE = 'SET_LOGIN_STATE'
export const SET_USER = 'SET_USER'
export const SET_AUTHENTICATING = 'SET_AUTHENTICATING'
export const DO_LOGOUT = 'DO_LOGOUT'
