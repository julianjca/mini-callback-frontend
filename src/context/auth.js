import React, { useEffect, createContext, useReducer } from 'react'
import cookie from 'react-cookies'
import axios from 'axios'

import { SET_LOGIN_STATE, SET_USER, SET_AUTHENTICATING } from '../constants'

const AuthStateContext = createContext()
const AuthDispatchContext = createContext()

const reducer = (state, action) => {
  switch (action.type) {
    case SET_LOGIN_STATE: {
      return { ...state, isLoggedIn: action.isLoggedIn }
    }

    case SET_USER: {
      return { ...state, user: action.user }
    }

    case SET_AUTHENTICATING: {
      return { ...state, isAuthenticating: action.isAuthenticating }
    }

    default:
      console.log('error')
  }
}

const AuthProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    isLoggedIn: false,
    user: null,
    isAuthenticating: true,
  })

  useEffect(() => {
    const cookies = cookie.load('userToken')
    const doAuth = async () => {
      dispatch({
        type: SET_AUTHENTICATING,
        isAuthenticating: true,
      })

      const res = await axios.get('http://localhost:3030/users/authenticate', {
        headers: {
          Authorization: `Bearer ${cookies}`,
        },
      })

      dispatch({
        type: SET_USER,
        user: res.data.user,
      })
      dispatch({
        type: SET_LOGIN_STATE,
        isLoggedIn: true,
      })
      dispatch({
        type: SET_AUTHENTICATING,
        isAuthenticating: false,
      })
    }
    if (typeof window !== undefined) {
      if (cookies) {
        doAuth()
      } else {
        cookie.remove('userToken')
        dispatch({
          type: SET_USER,
          user: null,
        })
        dispatch({
          type: SET_LOGIN_STATE,
          isLoggedIn: false,
        })
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>{children}</AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

const useAuthState = () => {
  const context = React.useContext(AuthStateContext)

  return context
}

const useAuthDispatch = () => {
  const context = React.useContext(AuthDispatchContext)

  return context
}

export { AuthProvider, useAuthState, useAuthDispatch }
