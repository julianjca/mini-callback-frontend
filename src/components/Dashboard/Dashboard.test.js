/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react'
import { render, fireEvent } from '../../../test-utils'
import * as context from '../../context/auth'
import * as nextRouter from 'next/router'
import { DO_LOGOUT } from '../../constants'

import Dashboard from './index'

jest.mock('../../context/auth')
jest.mock('next/router')
jest.mock('axios')

// eslint-disable-next-line no-import-assign
nextRouter.useRouter = jest.fn()
const push = jest.fn()
nextRouter.useRouter.mockImplementation(() => ({ push }))

const useAuthDispatch = jest.fn()
context.useAuthDispatch.mockImplementation(() => useAuthDispatch)

const callbacks = [
  {
    bank_code: 'BANK_ABC',
    businessId: '270cd61d-1965-4ed2-bbf3-a91425a030b8',
    callbackResponseCode: 200,
    createdAt: '2021-02-18T12:54:16.810Z',
    id: 'e8980029-04a0-4b9c-8c64-182c8a81cc43',
    timestamp: '2021-02-18T03:13:01.411Z',
    transaction_id: '123',
    updatedAt: '2021-02-18T12:54:16.810Z',
    virtual_account: '1234',
  },
]

describe('Dashboard', () => {
  test('renders the right text', () => {
    const element = render(<Dashboard callbacks={callbacks} />)

    expect(element).toMatchSnapshot()
  })

  test('runs log out function', () => {
    const element = render(<Dashboard callbacks={callbacks} />)

    const logOutButton = element.getByText('Log out')
    fireEvent.click(logOutButton)
    expect(useAuthDispatch).toHaveBeenCalledTimes(1)
    expect(useAuthDispatch).toHaveBeenCalledWith({
      type: DO_LOGOUT,
    })
  })

  test('renders the table', () => {
    const element = render(<Dashboard callbacks={callbacks} />)

    expect(element.getByText('BANK_ABC')).toBeInTheDocument()
    expect(element.getByText('200')).toBeInTheDocument()
    expect(element.getByText('123')).toBeInTheDocument()
    expect(element.getByText('1234')).toBeInTheDocument()
    expect(element.getByText('e8980029-04a0-4b9c-8c64-182c8a81cc43')).toBeInTheDocument()
  })

  test('open detail page when the button is clicked', () => {
    const element = render(<Dashboard callbacks={callbacks} />)

    const seeDetailButton = element.getByText('See detail')

    fireEvent.click(seeDetailButton)

    expect(push).toHaveBeenCalledTimes(1)
    expect(push).toHaveBeenCalledWith('/callbacks/[id]', `/callbacks/${callbacks[0].id}`, {
      shallow: true,
    })
  })
})
