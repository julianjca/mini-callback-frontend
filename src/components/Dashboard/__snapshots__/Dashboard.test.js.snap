// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders the right text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="chakra-ui-light"
  >
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; top: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; top: 0px; right: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto; bottom: 0px; right: 0px; left: 0px;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; bottom: 0px; right: 0px;"
      />
    </div>
    <div>
      <div
        class="css-bskjhk"
      >
        <div
          class="css-1lekzkb"
        >
          <div
            class="css-4g6ai3"
          >
            <h2
              class="chakra-heading css-zey6tx"
            >
              Callback UI
            </h2>
          </div>
          <div
            class="css-0"
          >
            <button
              class="chakra-button css-wnilqh"
              type="button"
            >
              Log out
            </button>
          </div>
        </div>
        <div
          class="css-46b038"
        >
          <button
            class="chakra-button css-myeq6i"
            type="button"
          >
            Send callback
          </button>
        </div>
        <table
          class="chakra-table css-93szjo"
          role="table"
        >
          <thead
            class="css-0"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1moe7jb"
              >
                Callback ID
              </th>
              <th
                class="css-1moe7jb"
              >
                Transaction ID
              </th>
              <th
                class="css-1moe7jb"
              >
                Virtual Account
              </th>
              <th
                class="css-1moe7jb"
              >
                Bank Code
              </th>
              <th
                class="css-1moe7jb"
              >
                Status
              </th>
              <th
                class="css-1moe7jb"
              >
                Detail
              </th>
            </tr>
          </thead>
          <tbody
            class="css-0"
          >
            <tr
              class="css-0"
              role="row"
            >
              <td
                class="css-hykeff"
                role="gridcell"
              >
                e8980029-04a0-4b9c-8c64-182c8a81cc43
              </td>
              <td
                class="css-hykeff"
                role="gridcell"
              >
                123
              </td>
              <td
                class="css-hykeff"
                role="gridcell"
              >
                1234
              </td>
              <td
                class="css-hykeff"
                role="gridcell"
              >
                BANK_ABC
              </td>
              <td
                class="css-hykeff"
                role="gridcell"
              >
                <span
                  class="chakra-badge css-tjtztc"
                >
                  200
                </span>
              </td>
              <td
                class="css-hykeff"
                role="gridcell"
              >
                <button
                  class="chakra-button css-1d11b2j"
                  type="button"
                >
                  See detail
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-bskjhk"
    >
      <div
        class="css-1lekzkb"
      >
        <div
          class="css-4g6ai3"
        >
          <h2
            class="chakra-heading css-zey6tx"
          >
            Callback UI
          </h2>
        </div>
        <div
          class="css-0"
        >
          <button
            class="chakra-button css-wnilqh"
            type="button"
          >
            Log out
          </button>
        </div>
      </div>
      <div
        class="css-46b038"
      >
        <button
          class="chakra-button css-myeq6i"
          type="button"
        >
          Send callback
        </button>
      </div>
      <table
        class="chakra-table css-93szjo"
        role="table"
      >
        <thead
          class="css-0"
        >
          <tr
            class="css-0"
            role="row"
          >
            <th
              class="css-1moe7jb"
            >
              Callback ID
            </th>
            <th
              class="css-1moe7jb"
            >
              Transaction ID
            </th>
            <th
              class="css-1moe7jb"
            >
              Virtual Account
            </th>
            <th
              class="css-1moe7jb"
            >
              Bank Code
            </th>
            <th
              class="css-1moe7jb"
            >
              Status
            </th>
            <th
              class="css-1moe7jb"
            >
              Detail
            </th>
          </tr>
        </thead>
        <tbody
          class="css-0"
        >
          <tr
            class="css-0"
            role="row"
          >
            <td
              class="css-hykeff"
              role="gridcell"
            >
              e8980029-04a0-4b9c-8c64-182c8a81cc43
            </td>
            <td
              class="css-hykeff"
              role="gridcell"
            >
              123
            </td>
            <td
              class="css-hykeff"
              role="gridcell"
            >
              1234
            </td>
            <td
              class="css-hykeff"
              role="gridcell"
            >
              BANK_ABC
            </td>
            <td
              class="css-hykeff"
              role="gridcell"
            >
              <span
                class="chakra-badge css-tjtztc"
              >
                200
              </span>
            </td>
            <td
              class="css-hykeff"
              role="gridcell"
            >
              <button
                class="chakra-button css-1d11b2j"
                type="button"
              >
                See detail
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
